tmk_core/common/chibios/eeprom_stm32.c

```diff
@@ -14,185 +14,793 @@
+#endif
+
+/* In-memory contents of emulated eeprom for faster access */
+/* *TODO: Implement page swapping */
+static uint16_t WordBuf[FEE_DENSITY_BYTES / 2];
+static uint8_t *DataBuf = (uint8_t *)WordBuf;
+
+/* Pointer to the first available slot within the write log */
+static uint16_t *empty_slot;
+
+#ifdef DEBUG_EEPROM
+void dumpEepromDataBuf(void) {
+    int empty_rows = 0;
+    for (uint16_t i = 0; i < FEE_DENSITY_BYTES; i++) {
+        if (i % 16 == 0) {
+            if (i >= FEE_DENSITY_BYTES - 16) {
+                /* Make sure we display the last row */
+                empty_rows = 0;
+            }
+            /* Check if this row is uninitialized */
+            ++empty_rows;
+            for (uint16_t j = 0; j < 16; j++) {
+                if (DataBuf[i + j]) {
+                    empty_rows = 0;
+                    break;
+                }
+            }
+            if (empty_rows > 1) {
+                /* Repeat empty row */
+                if (empty_rows == 2) {
+                    /* Only display the first repeat empty row */
+                    dprintf("*\n");
+                }
+                i += 15;
+                continue;
+            }
+            dprintf("%04x", i);
+        }
+        if (i % 8 == 0) dprintf(" ");
+
+        dprintf(" %02x", DataBuf[i]);
+        if ((i + 1) % 16 == 0) {
+            dprintf("\n");
+        }
+    }
+}
+#endif
 
-uint8_t DataBuf[FEE_PAGE_SIZE];
-/*****************************************************************************
- *  Delete Flash Space used for user Data, deletes the whole space between
- *  RW_PAGE_BASE_ADDRESS and the last uC Flash Page
- ******************************************************************************/
 uint16_t EEPROM_Init(void) {
-    // unlock flash
-    FLASH_Unlock();
+    /* Load emulated eeprom contents from compacted flash into memory */
+    uint16_t *src  = (uint16_t *)FEE_PAGE_BASE_ADDRESS;
+    uint16_t *dest = (uint16_t *)DataBuf;
+    for (; src < (uint16_t *)FEE_PAGE_LAST_ADDRESS; ++src, ++dest) {
+        *dest = ~*src;
+    }
 
-    // Clear Flags
-    // FLASH_ClearFlag(FLASH_SR_EOP|FLASH_SR_PGERR|FLASH_SR_WRPERR);
+#ifdef DEBUG_EEPROM
+    dprintf("EEPROM_Init Compacted Pages:\n");
+    dumpEepromDataBuf();
+    dprintf("EEPROM_Init Write Log:\n");
+#endif
+
+    /* Replay write log */
+    uint16_t *log_addr;
+    for (log_addr = (uint16_t *)FEE_WRITE_LOG_BASE_ADDRESS; log_addr < (uint16_t *)FEE_WRITE_LOG_LAST_ADDRESS; ++log_addr) {
+        uint16_t address = *log_addr;
+        if (address == FEE_EMPTY_WORD) {
+            break;
+        }
+        /* Check for lowest 128-bytes optimization */
+        if (!(address & FEE_WORD_ENCODING)) {
+            uint8_t bvalue = (uint8_t)address;
+            address >>= 8;
+            DataBuf[address] = bvalue;
+#ifdef DEBUG_EEPROM
+            dprintf("DataBuf[0x%02x] = 0x%02x;\n", address, bvalue);
+#endif
```

Comment:
Maybe a wrapper to avoid some of chatter in the file? There was a PR somewhere that did something similar, will add a link when i can find it.

Something like....
```c
#ifdef DEBUG_EEPROM
#    define eeprom_printf dprintf
#else
#    define eeprom_printf 
#endif
```

Reply:
I switched to using the eeprom_printf wrapper